cranelift/codegen/src/machinst/buffer.rs

```diff
@@ -451,36 +467,62 @@ impl<I: VCodeInst> MachBuffer<I> {
                 break;
             }
 
+            // For any branch, conditional or unconditional:
+            // - If the target is a label at the current offset, then remove
+            //   the conditional branch, and reset all labels that targetted
+            //   the current offset (end of branch) to the truncated
+            //   end-of-code.
+            if self.resolve_label_offset(b.target) == cur_off {
+                trace!("branch with target == cur off; truncating");
+                self.truncate_last_branch();
+                continue;
+            }
+
             // If latest is an unconditional branch:
             // - For each label at this point, make the label an alias of
             //   the branch target. We can now assume below that the
             //   unconditional branch is reachable only via fallthrough, and we
             //   are free to remove it in an optimization.
             // - If there is a prior unconditional branch that ends just before
             //   this one begins, then we can truncate this branch, because it is
-            //   entirely unreachable (due to above). Trim the end of the
-            //   `labels_by_offset` array and continue around the loop.
+            //   entirely unreachable (due to above). Do so and continue around
```

Comment:
This doesn't entail any kind of fixpointing behaviour, right?

Reply:
Not in the usual sense, no. On each iteration of the loop we munch a branch (or redirect its labels elsewhere); we don't ever revisit a branch.